client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::{CurrentSpan, filter::{EnvFilter, Directive}, layer::SubscriberExt};
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Estimate of the max size base RPC payload when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload does not
+// include the RPC result.
+const BASE_RPC_PAYLOAD: usize = 500;
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &'static str = "";
+// The name of a field required for all events.
+const REQUIRED_EVENT_FIELD: &'static str  = "method";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				if metadata.is_span() {
+					return true;
+				} else if metadata.fields().field(REQUIRED_EVENT_FIELD).is_some() {
+					// it is an event, so we check if has the method field
+					return true;
+				}
```

Comment:
This is the more fine grained filter for events.
@mattrutherford & @dvdplm let me know what you think.